ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -5,6 +5,91 @@
 #include "ruby_helpers.h"
 #include <errno.h>
 
+#define MAX_FRAMES_LIMIT 10000
+
+// A compact representation of a stacktrace frame for a heap allocation.
+typedef struct {
+  char *name;
+  char *filename;
```

Comment:
I did try this first but it required casting out of a const char*? Seemed like bad hygiene xD

Reply:
Hmmm I'm not sure I follow why it needed a cast?

To clarify, I was thinking we may be able to keep this as

```c
typedef struct {
  ddog_CharSlice name;
  ddog_CharSlice filename;
  int64_t line;
} heap_frame;
```

which hopefully might simplify some of the comparisons and conversions later (e.g. no more having to support both charslice and char* and whatnot).

Reply:
The cast is to be able to fill allocate a `ddog_CharSlice` dynamically which we'll then need to free later:

<img width="515" alt="image" src="https://github.com/DataDog/dd-trace-rb/assets/373323/173919ba-d82b-4645-9412-ff80f17467bc">

We can do it but feels dirtier 😅

Reply:
Oh, I see. That is super annoying... I guess I've been avoiding all this time since I never malloc'd something to be used as a `CharSlice`, it always came from somewhere else.

On the other hand, it seems... annoying as well to change the libdatadog type to not be const, since in a lot of cases it probably is the right choice. :(